### DISCUSSION TITLE
use on_failure_callback or on_success_callback error, failure_callback() missing 1 required positional argument: 'context'; 27572)

### DISCUSSION THREAD

**Discussion by legend-xu** [apache/airflow#25052] [Q&A] [answered]
def failure_callback(context):
    dag_id = context['dag'].dag_id
    task_id = context['dag'].task_id
    reason = context['exception']
    mail_msg = f"""
        <p>日期:{datetime.now().strftime("%Y-%m-%d %H:%M:%S")}</p>
        <p>DAG:{dag_id}...</p>
        <p>Task:{task_id}</p>
        <p>reason:{reason}</p>
        """
    send_email_fun(mail_msg)

default_args = {
    'owner': 'airflow',
    'depends_on_past': False,
    'start_date': datetime(2022, 7, 14, 14, 30, 0),
    'email': ['test123@163.com'],
    'email_on_failure': False,
    'email_on_retry': False,
    'on_failure_callback': failure_callback,
    'on_success_callback': failure_callback,
}

dag = DAG(
    dag_id="test_send",
    default_args=default_args,
    schedule_interval=timedelta(days=1),
)

def print_hello():
    print("hello")

send_test = PythonOperator(
    task_id='send_email',
    # provide_context=True,
    python_callable=print_hello,
    dag=dag,
)

send_test

when I run, code error,failure_callback() missing 1 required positional argument: 'context'; 27572)
please help me , thank you!

**Comment by potiuk** [comment, marked as answer]
Your "print_hello" need "context" argument because callbacks are always handled with "context" parameter.

**Reply by legend-xu** [reply]
thanks!
